### PR TITLE
[WIP] Fix small issue with layers._cosine_dist

### LINKED ISSUE #2022: Cosine similarity calculation
I believe the function computing cosine similarity, found in layers.py, does not compute the denominator correctly if y is a matrix instead of a single vector:

** Current code **

```
def _cosine_dist(x, y):
  denom = (backend.sqrt(backend.sum(tf.square(x)) * backend.sum(tf.square(y))) + backend.epsilon())
  return backend.dot(x, tf.transpose(y)) / denom
```

** Possible fix **

```
def _cosine_dist(x, y):
  denom = backend.sqrt(backend.dot(tf.reshape(backend.sum(tf.square(x), axis=1), (-1, 1)), tf.reshape(backend.sum(tf.square(y), axis=1), (1, -1))) + backend.epsilon())
  
  return tf.math.divide(backend.dot(x, tf.transpose(y)), denom)
```

The denominator should not be equal for every computed similarity value. It should be specific for each pair of vectors for which we wish to compute similarity.

Edit by @rbharath: Fixed formatting on code samples

### PR DESCRIPTION
**Pull request by shak360**
Fixes #2022 

1. I think the previous code in `deepchem/models/layers.py` for `_cosine_dist` assumes that the input are vectors. This change should allow for this distance to work with tensors as inputs.

2. Added a unit test to `deepchem/models/tests/test_layers.py` to assert that this function is working properly

### CODE REVIEW

File: deepchem/models/tests/test_layers.py [hunk 1]
```diff
@@ -7,6 +7,14 @@
...
 class TestLayers(test_util.TensorFlowTestCase):
 
+  def test_cosine_dist(self):
+    """Test invoking _cosine_dist."""
+    x = np.ones((5, 4)).astype(np.float32)
+    y_same = np.ones((5, 4)).astype(np.float32)
+    y_far = np.zeros((5, 4)).astype(np.float32)
+    assert sum(layers._cosine_dist(x, y_same)) == 0
+    assert sum(layers._cosine_dist(x, y_far)) == 5
+
```
Comment (shak360) [inline deepchem/models/tests/test_layers.py:17]: Added unit test.

File: deepchem/models/layers.py [hunk 2]
```diff
@@ -429,9 +427,10 @@ def _cosine_dist(x, y):
   y: tf.Tensor
     Input Tensor
   """
-  denom = (backend.sqrt(backend.sum(tf.square(x)) * backend.sum(tf.square(y))) +
-           backend.epsilon())
-  return backend.dot(x, tf.transpose(y)) / denom
+  x_norm = tf.nn.l2_normalize(x, axis=1)
+  y_norm = tf.nn.l2_normalize(y, axis=1)
+  # the cosine distance is 1 - the cosine similarity
+  return 1. - tf.reduce_sum(tf.multiply(x_norm, y_norm), axis=1)
```
Comment (shak360) [inline deepchem/models/layers.py:433]: Modified cosine distance function.

File: deepchem/models/layers.py [hunk 3]
```diff
@@ -429,9 +427,9 @@ def _cosine_dist(x, y):
   y: tf.Tensor
     Input Tensor
   """
-  denom = (backend.sqrt(backend.sum(tf.square(x)) * backend.sum(tf.square(y))) +
-           backend.epsilon())
-  return backend.dot(x, tf.transpose(y)) / denom
+  x_norm = tf.nn.l2_normalize(x, axis=1)
+  y_norm = tf.nn.l2_normalize(y, axis=1)
+  return 1. - backend.dot(x_norm, tf.transpose(y_norm))
```
Comment (shak360) [inline deepchem/models/layers.py:432]: Modified cosine distance function.

File: deepchem/models/tests/test_layers.py [hunk 4]
```diff
@@ -7,6 +7,16 @@
...
+  def test_cosine_dist(self):
+    """Test invoking _cosine_dist."""
+    x = tf.ones((5, 4), dtype=tf.dtypes.float32, name=None)
+    y_same = tf.ones((5, 4), dtype=tf.dtypes.float32, name=None)
+    y_far = -1. * tf.ones((5, 4), dtype=tf.dtypes.float32, name=None)
+    close_cosine_dist = layers._cosine_dist(x, y_same)
+    far_close_dist = layers._cosine_dist(x, y_far)
+    assert tf.reduce_sum(close_cosine_dist) == 0.
+    assert tf.reduce_sum(far_close_dist) == 2. * 5. * 5.
+
```
Comment (shak360) [inline deepchem/models/tests/test_layers.py:19]: Added unit test.

File: deepchem/models/layers.py (hunk 3 above)
Comment (rbharath) [reply deepchem/models/layers.py:432]: I think this should be `backend.dot(x_norm, tf.transpose(y_norm))` instead of `1 - backend.dot(x_norm, tf.transpose(y_norm))` to compute the cosine distance right?

File: deepchem/models/layers.py [hunk 5]
```diff
@@ -429,9 +427,9 @@ def _cosine_dist(x, y):
   y: tf.Tensor
     Input Tensor
```
Comment (rbharath) [inline deepchem/models/layers.py:428]: Can you add details to the docstring with the types of the tensors being processed and details about the tensor being returned and add a usage example to the docstring?

File: deepchem/models/layers.py (hunk 3 above)
Comment (shak360) [reply deepchem/models/layers.py:432]: Yea I think I was confused about whether we want the cosine similarity or the cosine distance... I think if the goal is to return the inner product of the normalized inputs then you're right, it shouldn't be `1 - backend.dot(x_norm, tf.transpose(y_norm))` but rather just `backend.dot(x_norm, tf.transpose(y_norm)`.

File: deepchem/models/layers.py [hunk 6]
```diff
@@ -471,18 +471,83 @@ def call(self, inputs):
...
+
+  Returns
+  -------
+  tf.Tensor
+    Returns a tensor of shape (n, m), that is, n rows by m columns. 
+    Each i,j-th entry of this output tensor is the inner product between
+    the l2-normalized i-th row of the input tensor x and the
+    the l2-normalized j-th row of the output tensor y.
+
   """
```
Comment (shak360) [inline deepchem/models/layers.py:547]: Added a more thorough docstring with some examples.

File: deepchem/models/layers.py [hunk 7]
```diff
@@ -471,18 +471,83 @@ def call(self, inputs):
...
-  denom = (backend.sqrt(backend.sum(tf.square(x)) * backend.sum(tf.square(y))) +
-           backend.epsilon())
-  return backend.dot(x, tf.transpose(y)) / denom
+  x_norm = tf.math.l2_normalize(x, axis=1)
+  y_norm = tf.math.l2_normalize(y, axis=1)
+  return backend.dot(x_norm, tf.transpose(y_norm))
```
Comment (shak360) [inline deepchem/models/layers.py:550]: Returns the inner product of the l2-normalized rows of the input tensors.

File: deepchem/models/layers.py [hunk 8]
```diff
@@ -471,18 +471,83 @@ def call(self, inputs):
...
+  The cosine similarity between two equivalent vectors will be 1. The cosine
+  similarity between two equivalent tensors (input tensors where the elements are
+  the same), will be a tensor of ones. In this scenario, if the input tensors
+  a and y were each of shape (n,p), where each element in x and y were the same, then 
+  the output tensor would be a tensor of shape (n,n) with 1 in every entry.
+  The cosine similarity between two orthogonal vectors will be 0 (by definition).
+  If every row in x is orthogonal to every row in y, then the output will be a tensor
+  of 0s.
+
+  ```python
```
Comment (rbharath) [inline deepchem/models/layers.py:504]: Could you swap this over to doctest style? So this would look like

```
>>> import tensorflow as tf
>>> import deepchem.models.layers as layers
...
```
That way, the CI can run the code in the docstrings to make sure it remains correct

File: deepchem/models/layers.py [hunk 9]
```diff
@@ -471,18 +471,83 @@ def call(self, inputs):
 
 
 def _cosine_dist(x, y):
-  """Computes the inner product (cosine distance) between two tensors.
+  """Computes the inner product (cosine similarity) between two tensors.
```
Comment (rbharath) [inline deepchem/models/layers.py:474]: This docstring is great! Given the improved docs, we might as well go ahead and make this function public 
```
def cosine_dist(x, y)
```
Could you also add the public function to `deepchem/docs/layers.rst`? This should look like
```
.. autofunction:: deepchem.models.layers.cosine_dist
```

File: docs/layers.rst [hunk 10]
```diff
@@ -99,3 +99,5 @@ another tensor. DeepChem maintains an extensive collection of layers which perfo
 
 .. autoclass:: deepchem.models.layers.SetGather
   :members:
+
+.. autofunction:: deepchem.models.layers._cosine_dist
```
Comment (shak360) [inline docs/layers.rst:103]: Added function docstring to documentation (I think is what this does...)

File: deepchem/models/layers.py [hunk 11]
```diff
@@ -471,18 +471,86 @@ def call(self, inputs):
...
+  >>> x1 = identity_tensor[0:256,:]
+  >>> x2 = identity_tensor[256:512,:]
+  >>> # each row in x1 is orthogonal to each row in x2
+  >>> # the pairwise inner product of the rows in x and y will always be 0
+  >>> # the output tensor will be of shape (256,256)
+  >>> cos_sim_orth = layers._cosine_dist(x1,x2)
+  >>> assert tf.reduce_sum(cos_sim_orth) == 0 # True
+  True
+  >>> assert all([cos_sim_orth.shape[dim] == 256 for dim in range(2)]) # True
+  True
```
Comment (shak360) [inline deepchem/models/layers.py:529]: Wrote this in doctest format; I didn't know this was actually possible and is kind of a cool feature

File: deepchem/models/layers.py [hunk 12]
```diff
@@ -471,18 +471,87 @@ def call(self, inputs):
...
+  the same), will be a tensor of 1s. In this scenario, if the input tensors
+  `x` and `y` were each of shape `(n,p)`, where each element in `x` and `y`
+  were the same, then the output tensor would be a tensor of shape `(n,n)` 
+  with 1 in every entry.
+  
+  >>> import tensorflow as tf
+  >>> import deepchem.models.layers as layers
+  >>> x = tf.ones((6, 4), dtype=tf.dtypes.float32, name=None)
+  >>> y_same = tf.ones((6, 4), dtype=tf.dtypes.float32, name=None)
+  >>> # x and y are the same tensor (equivalent at every element)
```
Comment (rbharath) [inline deepchem/models/layers.py:506]: Instead of having theses as comments, you can break out a little paragraph of text. That might make it easier for readers to follow along

File: docs/layers.rst (hunk 10 above)
Comment (rbharath) [reply docs/layers.rst:103]: Yep this works! However, we should rename `_cosine_dist` to `cosine_dist` (in general having a `_` at the front indicates a function is private and not to be used more broadly). Since the implementation is cleaned up, we should make it public so more people can use it :)

File: deepchem/models/layers.py (hunk 12 above)
Comment (shak360) [reply deepchem/models/layers.py]: Sounds good, this should be changed now : )

File: docs/layers.rst (hunk 10 above)
Comment (shak360) [reply docs/layers.rst]: OK, I changed everywhere it's called to be `cosine_dist` at least in the `layers.py` file. Let me know if there are other locations that need to be changed.

**Comment by shak360** [comment]
Thanks @rbharath : ), I'm learning a lot. I had a couple of questions, mainly about administrative stuff, which you can answer here or in the call on Friday. I was just wondering about etiquette, in terms of these pull requests. I know that I should only submit pull requests when I'm near done or close to finishing (which I guess in that case I put a `[WIP]`, in front of the title right?). If it's the case where I want to see the output of Travis CI, just to see if everything is working so far, but not necessarily claim that it's ready for a review, how should I go about doing that? Also, should I be manually requesting you for reviews, or do you have a personal schedule where you go through all pending pull requests (I just don't wanna seem annoying if I keep asking you for a request manually if you already have a plan for going through them)?

Also as a next issue, I was possibly hoping to tackle something like #2014 (Keras layers unit tests) or #2037 (Performance curves). Let me know what you think, or if you think I should approach something else.

**Comment by peastman** [comment]
As long as you mark the PR with [WIP], it's fine to submit it at any stage.  Sometimes people do that to get early feedback on the approach they're taking before spending too much time on it.  Just post a comment when you think it's ready for review.  And that doesn't mean it has to be finished.  Feel free to ask people to look it over at any point where you think it would be useful.

**Comment by rbharath** [comment]
+1 to @peastman's comments!

Both #2014 or #2037 would be great next contributions :). Once you're up for it, I'd recommend that another great place to contribute would be to contribute a new model class to DeepChem. There have been a lot of really neat models that have come out in the last year or two that we should add in
